### PR TITLE
Use least-qualified names for conflicting class names

### PR DESCRIPTION
**Pull request by auke-**
After the quick fix for conflicting names (#320) this is my first shot at making name resolution more straight forward.

There are a couple of simple scenarios:
- same class
Always use the simple name, e.g. x.y.A referenced from x.y.A resolves to A and x.y.A.B referenced from x.y.A.B resolves to B.

- parent class
Always use the simple name, e.g. x.y.A.B referenced from x.y.A.B.C resolves to B and  x.y.A referenced from x.y.A.B.C resolves to A.

- nested class
Always use the qualified name starting at the nesting, e.g. x.y.A.B referenced from x.y.A resolves to B and x.y.A.B.C referenced from x.y.A resolves to B.C.

And some complex scenarios:
- nested class of a parent class
Use the qualified name starting at the nesting, e.g. x.y.A.D referenced from x.y.A.B.C resolves to D.

- other class in the same package
Use the qualified name, e.g. x.y.B.C referenced from x.y.A. resolves to B.C.

- other class in another package
Use the qualified name and import top-level name, e.g. x.y.z.B.C referenced from x.y.A. resolves to B.C and imports x.y.z.B.

If (in the complex scenarios) the top-level name of the resolved name conflicts with another name (same class, parent, member, member of a parent, imported class or referenced class) the name of the enclosing class is used. If the top-level class conflicts, the fully qualified name is used (package and class names).

There are some issues with my current approach:
- Information about inherited classes and interfaces isn't available (see #77).
- Type information of annotated classes and classes implementing or extending classes isn't available on the type stack (should this be changed in TypeSpec.emit?).
- I'm not sure if the methods nestedName and isNestedClass should be methods of ClassName or CodeWriter.
- Not all scenarios are covered by tests.
- Scenarios with anonymous nested classes aren't covered by tests.

### CODE REVIEW

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 1]
```diff
@@ -272,106 +273,158 @@ private void emitLiteral(Object o) throws IOException {
...
   String lookupName(ClassName className) {
-    // Different package than current? Just look for an import.
-    if (!className.packageName().equals(packageName)) {
-      if (conflictsWithLocalName(className)) {
-        return className.toString(); // A local name conflicts? Use the fully-qualified name.
+
+    // Create ClassName from current stack
+    ClassName stackContext = null;
+    for (TypeSpec typeSpec : typeSpecStack) {
+        if (typeSpec.name == null) {
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: can you reformat this with two-space indentation?

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 2]
```diff
@@ -272,106 +273,158 @@ private void emitLiteral(Object o) throws IOException {
    * names visible due to inheritance.
    */
   String lookupName(ClassName className) {
-    // Different package than current? Just look for an import.
-    if (!className.packageName().equals(packageName)) {
-      if (conflictsWithLocalName(className)) {
-        return className.toString(); // A local name conflicts? Use the fully-qualified name.
+
+    // Create ClassName from current stack
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Just as a heads up, in general, [comments should be sentences][1]. All your comments are missing periods. Seems minor, I know, but the linked article points out some compelling arguments for it. Plus it help reinforce good commenting habit.

 [1]: nedbatchelder.com/blog/201401/comments_should_be_sentences.html

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 3]
```diff
@@ -272,106 +273,158 @@ private void emitLiteral(Object o) throws IOException {
...
+        } else {
+            stackContext = stackContext.nestedClass(typeSpec.name);
+        }
+    }
+
+    if (stackContext == null) {
+      // There is no stack context, e.g. annotation or interface on top-level classes
+      if (className.packageName().equals(packageName)) {
+        return Util.join(".", className.simpleNames());
+      } else {
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: You don't need `else` after `return`, and it makes the presence of this comment not seem so strange.
```java
if (foo) {
  return "foo";
}
// There's no way to know!
```

**Comment by swankjesse** [comment]
Is there a simpler way to implement this by working backwards?

Start with the fully qualified name `A.B.C.D`.
Drop the prefix and attempt to resolve `B.C.D`.
Does that resolve to `A.B.C.D` ? If no, return the full path. If yes, repeat the process.

**Comment by swankjesse** [comment]
(we'd need a `resolve()` function that would take a partial name and return a full name)

**Comment by auke-** [comment]
The simple scenarios (same class, nested class or parent class) are relatively easy to detect and handled, because the names can't conflict with another name in the same context.

For the other scenarios there are two issues: what is the shortest relative reference and does it conflict with another name in the current context.

I don't see how working backwards makes it easier, but that more because I don't  see how that would work.

**Comment by auke-** [comment]
I've rebased the commit onto master. I'm looking for a way to get this merged. Can you suggest how I should split this up in smaller parts, make the logic more straightforward or change this commit in another way to make it easier to review and merge? Thanks

**Comment by swankjesse** [comment]
I want to make an attempt at a resolution-based implementation to see if what I'm imagining will actually work. If it doesn't (it very likely won't!) I'll merge this! Sorry for the hold up.

**Comment by auke-** [comment]
No problem. Note that there is (at least) one issue with my patch:
* Type information of annotated classes and classes implementing or extending classes isn't available on the type stack

This means that there is no way to detect name conflicts for interfaces:
```
package foo;
class Collection implements java.util.Collection {
...
}
```

Or annotations:
```
package foo;
@java.lang.FunctionalInterface
interface FunctionalInterface {
...
}
```

This is only an issue for top-level classes (not an unusual case) and could be addressed in TypeSpec.emit (invoke codeWriter.pushType before writing annotations and interfaces).

**Comment by swankjesse** [comment]
@auke- here's resolution-based importing. https://github.com/square/javapoet/pull/331

**Comment by auke-** [comment]
I haven't had time to look into your solution until now and it seems you fixed the issue. The latest snapshot version generates a lot cleaner imports and class-references on our codebase. Thanks!

**Pull request closed without merging**